### PR TITLE
WIP: Enable sparse checkouts for referenced repositories

### PR DESCRIPTION
**Pull request by xenji**
Dear @jacebrowning ,

thanks a lot for building this tool! I highly appreciate that you solved one of my big current problems for me :D

Hereby I propose an addition to gitman, which enables sparse checkouts of the referenced repositories. The PR is **incomplete**, because tests and documentation are still missing for the addition. I wanted to get your feedback as early as possible, but also wanted to have something real to talk about. My python skills are limited, so please bare with me in regards to the chosen implementation.

## What is the value added with this PR?
I use gitman in order to pull together protobuf definitions from different repositories. Most of the remotes do not only carry the protobufs, but also the application that is using it. In order to reduce the clutter in the vendor folder, I'd like to introduce the option to select the paths that are relevant for me. In addition, I frequently work on bigger monorepos, which makes it even more valuable for me.

Before I continue my efforts, I'd like to have your feedback on it and I might need some help setting up the environment correctly. I've managed to test my changes locally, but using the Makefile as stated in the contributors guide leads me to an error related to a missing dependency when attempting to execute the tests.

Cheers!
Mario

Here is a sample log from a successful run:
```
mmueller at Brick in ~/Development/src/gitman-test
$ gitman install -v
[INFO    ] Installing dependencies: <all>
[INFO    ] Defaulting to locked sources...
[INFO    ] $ mkdir -p /Users/mmueller/Development/src/gitman-test/vendor
[INFO    ] $ cd /Users/mmueller/Development/src/gitman-test/vendor
[INFO    ] Updating source files...
[INFO    ] $ git -C sample_dependency init
[INFO    ] $ git -C sample_dependency config core.sparseCheckout true
[INFO    ] $ git -C sample_dependency remote add -f origin /Users/mmueller/.gitcache/hellogitworld.reference
[INFO    ] $ git -C sample_dependency pull origin HEAD
[INFO    ] $ cd sample_dependency
[INFO    ] $ git remote set-url origin https://github.com/githubtraining/hellogitworld
[INFO    ] $ git fetch --tags --force --prune origin
[INFO    ] $ git checkout --force ebbbf773431ba07510251bb03f9525c7bab2b13a
[INFO    ] Defaulting to locked sources...
[INFO    ] $ cd /Users/mmueller/Development/src/gitman-test/vendor
[INFO    ] Running install scripts...
[INFO    ] $ cd sample_dependency
[INFO    ] (no scripts to run)
[INFO    ] Installed 1 dependency
```

### CODE REVIEW

**Comment by jacebrowning** [comment]
I can see this feature being useful. Let's get your environment working first.

> but using the Makefile as stated in the contributors guide leads me to an error related to a missing dependency when attempting to execute the tests

What is the error? Is anything missing when you run `$ make doctor`?

**Comment by xenji** [comment]
Let me give you a quick overview:
```
$ pwd
/Users/mmueller/Development/src/github.com/gitman

$ pyenv global
3.6.4

$ pyenv local
pyenv: no local version configured for this directory
```
Here is the doctor output:
```
$ make doctor
bin/verchew

Checking for Make...

$ make --version
GNU Make 3.81
✔ MATCHED: GNU Make

Checking for Python...

$ python --version
Python 3.6.4
✔ MATCHED: Python 3.6.

Checking for pipenv...

$ pipenv --version
pipenv, version 8.3.0
✔ MATCHED: 8.3.

Checking for Git...

$ git --version
git version 2.16.1
✔ MATCHED: 2.

Checking for pandoc...

$ pandoc --version
sh: command not found: pandoc
✘ EXPECTED: 1.
⭑ MESSAGE: This is only needed to generate the README for PyPI.

Checking for Graphviz...

$ dot -V
sh: command not found: dot
⚠ EXPECTED: 2.
⭑ MESSAGE: This is only needed to generate UML diagrams for documentation.

Results: ✔ ✔ ✔ ✔ ✘ ⚠
```

Here is the `make test` output that I am stuck with:
```
$ make test
pipenv run py.test -ra -vv --random --random-seed=1517076713 --cov=gitman --no-cov-on-fail --cov-report=term-missing:skip-covered --cov-report=html gitman tests --junitxml=xmlreport/overall.xml
Traceback (most recent call last):
  File "/Users/mmueller/Development/src/github.com/gitman/.venv/lib/python3.5/site-packages/_pytest/config.py", line 325, in _getconftestmodules
    return self._path2confmods[path]
KeyError: local('/Users/mmueller/Development/src/github.com/gitman/gitman/tests')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/mmueller/Development/src/github.com/gitman/.venv/lib/python3.5/site-packages/_pytest/config.py", line 356, in _importconftest
    return self._conftestpath2mod[conftestpath]
KeyError: local('/Users/mmueller/Development/src/github.com/gitman/gitman/tests/conftest.py')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/mmueller/Development/src/github.com/gitman/.venv/lib/python3.5/site-packages/_pytest/config.py", line 362, in _importconftest
    mod = conftestpath.pyimport()
  File "/Users/mmueller/Development/src/github.com/gitman/.venv/lib/python3.5/site-packages/py/_path/local.py", line 668, in pyimport
    __import__(modname)
  File "/Users/mmueller/Development/src/github.com/gitman/.venv/lib/python3.5/site-packages/_pytest/assertion/rewrite.py", line 216, in load_module
    py.builtin.exec_(co, mod.__dict__)
  File "/Users/mmueller/Development/src/github.com/gitman/gitman/tests/conftest.py", line 7, in <module>
    import yorm
ImportError: No module named 'yorm'
ERROR: could not load /Users/mmueller/Development/src/github.com/gitman/gitman/tests/conftest.py

make: *** [test-all] Error 4
```

**Comment by jacebrowning** [comment]
That's odd. Maybe something strange happened when the virtual environment was created. Try:

```
$ make clean-all
$ make test
```

You should see the project's dependencies get installed.

**Comment by xenji** [comment]
Wohoo. The clean all did the trick. 

`5 failed, 155 passed, 4 skipped in 95.46 seconds`

I go ahead and add tests and documentation for it. I'm no native english speaker, please give the grammar and spelling some extra attention in the docs later.

**Comment by xenji** [comment]
Hm, strange. I did fix the existing tests so far, at least on my system and on travis. AppVeyor seems to have a different issue here: https://ci.appveyor.com/project/jacebrowning/gitman/build/1.0.217/job/h38e7v0tvkuyhj6a

I'll add new tests for the feature anyway.

**Comment by jacebrowning** [comment]
Sounds good! AppVeyor is also broken on `develop` so I will need to fix it there.

File: gitman/models/source.py [hunk 1]
```diff
@@ -16,19 +16,21 @@
 @yorm.attr(rev=String)
 @yorm.attr(link=NullableString)
 @yorm.attr(scripts=List.of_type(String))
+@yorm.attr(sparse_paths=List.of_type(String))
```
Comment (jacebrowning) [inline gitman/models/source.py]: The order of these decorators determines the order of attributes in generated YAML files. I'd prefer the `sparse_paths` setting to be between `repo` and `rev`.

Comment (xenji) [reply gitman/models/source.py]: Allright. Having `scripts` as last option makes sense, because it can grow quite large.

Comment (xenji) [reply gitman/models/source.py]: Changed in the latest commit.

**Comment by xenji** [comment]
Updating the demo case left me with some work to do. The demo gitman.yml refers multiple times to the same repo and a sparse checkout seems to have an unintended side-effect (see travis build). I need to investigate this further before I continue with the tests.

File: README.md [hunk 2]
```diff
@@ -54,6 +54,11 @@ sources:
   rev: master
   scripts:
   - chmod a+x truffleHog.py
+- name: fontawesome
+  repo: https://github.com/FortAwesome/Font-Awesome
+  rev: master
+  sparse_paths:
```
Comment (jacebrowning) [inline README.md]: Let's move this up a line to match the generated configuration.

File: README.md [hunk 3]
```diff
@@ -83,9 +88,10 @@ which will essentially:
 1. Create a working tree at `<root>`/`<location>`/`<name>`
 2. Fetch from `repo` and checkout the specified `rev`
 3. Symbolically link each `<location>`/`<name>` from `<root>`/`<link>` (if specified)
-4. Repeat for all nested working trees containing a config file
-5. Record the actual commit SHAs that were checked out (with `--lock` option)
-6. Run optional post-install scripts for each dependency
+4. Materialize only the selected paths from `sparse_paths`, if present in the config file
+5. Repeat for all nested working trees containing a config file
+6. Record the actual commit SHAs that were checked out (with `--lock` option)
+7. Run optional post-install scripts for each dependency
```
Comment (jacebrowning) [inline README.md]: This content is good, but I'm not sure if it's applicable to most users. This might make more sense in it's own `use-cases` page.

Comment (xenji) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (xenji) [reply README.md]: done

**Comment by xenji** [comment]
I found the issue from the travis build. I resolved it by passing the correct revision to the clone operation. Git errors on none of the sparse paths available in the HEAD revision. See e9810fe7363a21eed3febd48104b99d8b69dd515 for details.

**Comment by xenji** [comment]
Seems like I accidentally fixed the appveyor check... somehow.
Tests are in, but I cannot oversee if the coverage and cases are sufficient. Could you please have a look and give me feedback? Thanks!

**Comment by jacebrowning** [comment]
This looks pretty good to me. You'll want to add a line to `mkdocs.yml` to get your new documentation page to show up. Run `$ make mkdocs-live` to see what this will look like.

**Comment by xenji** [comment]
Docs look good to me, link is in the right place, changelog updated.

**Comment by jacebrowning** [comment]
Give this a try with `$ pip install gitman==1.5b2`.

**Pull request merged**